### ISSUE TITLE
koreader crash when highlight text in pdf

### ISSUE THREAD

**Issue by tlhc** [koreader/koreader#2954]
* KOReader version: v2015.11-1054
* Device: KPW3 

1.OTA to last version 1054
2.open pdf and select text 
3.crash..
here is the logs:

---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 [*] Current time: 06/12/17-23:28:58

06/12/17-23:28:59 INFO  initializing for device KindlePaperWhite3
06/12/17-23:28:59 INFO  framebuffer resolution: {
    ["h"] = 1448,
    ["w"] = 1072
}
[ko-input] Forked off fake event generator(pid:6762).
06/12/17-23:29:00 INFO  opening file /mnt/us/documents/LKMPG/lkmpg.pdf
06/12/17-23:29:00 INFO  Loading plugins from directory: plugins
06/12/17-23:29:01 INFO  Plugin  plugins/hello.koplugin/main.lua  has been disabled.
06/12/17-23:29:01 INFO  RD loaded plugin batterystat at plugins/batterystat.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin calibrecompanion at plugins/calibrecompanion.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin evernote at plugins/evernote.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin goodreads at plugins/goodreads.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin keepalive at plugins/keepalive.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin kobolight at plugins/kobolight.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin kosync at plugins/kosync.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin newsdownloader at plugins/newsdownloader.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin percepton_expander at plugins/perceptionexpander.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin readtimer at plugins/readtimer.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin statistics at plugins/statistics.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin systemstat at plugins/systemstat.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin terminal at plugins/terminal.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin timesync at plugins/timesync.koplugin
06/12/17-23:29:01 INFO  RD loaded plugin zsync at plugins/zsync.koplugin
06/12/17-23:29:01 INFO  setting zoom mode to page
06/12/17-23:29:20 WARN  cannot find zsync file 404
./luajit: frontend/document/koptinterface.lua:723: attempt to index local 'box' (a nil value)
stack traceback:
	frontend/document/koptinterface.lua:723: in function 'inside_box'
	frontend/document/koptinterface.lua:730: in function 'box_distance'
	frontend/document/koptinterface.lua:743: in function 'getWordBoxIndices'
	frontend/document/koptinterface.lua:779: in function 'getTextFromBoxes'
	frontend/document/koptinterface.lua:996: in function 'getTextFromPositions'
	frontend/apps/reader/modules/readerhighlight.lua:280: in function 'handler'
	frontend/ui/widget/container/inputcontainer.lua:197: in function 'handleEvent'
	frontend/ui/uimanager.lua:446: in function 'sendEvent'
	frontend/ui/uimanager.lua:42: in function '__default__'
	frontend/ui/uimanager.lua:722: in function 'handleInput'
	frontend/ui/uimanager.lua:767: in function 'run'
	./reader.lua:193: in main chunk
	[C]: at 0x0000bef1
lipc-wait-event exited normally with status: 0

**Comment by Frenzie** [comment]
Is the PDF something that can be shared?

**Comment by tlhc** [comment]
sure,  http://www.tldp.org/LDP/lkmpg/2.6/lkmpg.pdf

**Comment by Frenzie** [comment]
I can't reproduce that issue on the emulator. Is it just any word? I tried a few from the first few pages at random.

**Comment by tlhc** [comment]
not any word, I think is every word,  when i upgrade to the last version. reopen koreader and highlight some text, it's crash, do you need my koreader config or something ?

**Comment by Frenzie** [comment]
Do you use any special settings? I've tried a bunch of things like crop and no-crop, scroll and no scroll, reflow and no reflow. I suppose it could be something Kindle-specific but hopefully it's not. :-P

**Comment by tlhc** [comment]
just reflow and always save documents and manual crop.
but the log is:  frontend/document/koptinterface.lua:723: attempt to index local 'box' (a nil value)
should we add some condition before use the box object or something ? may be just my kindle crash.. :),

**Comment by tlhc** [comment]
when I download the koreader-nightly-20170430, reinstall the koreader and attempt to OTA again，
koreader show already update to last version   v2015.11-1054 , it is normal ?

**Comment by Frenzie** [comment]
That shouldn't be an issue though. On pages without text it comes up with nothing (without crashing). On pages with text it should simply find the nearest textbox. With reflow it's sometimes slightly weird about some specifics and sometimes it doesn't find any textbox but it's still not crashing (obviously nor should it :-P).

I tested with `./kodev run -h=1448 -w=1072` (the resolution of your device) coupled with reflow on.

Screenshot of what I presume to be a textbox that's not scaling along:
![screenshot_2017-06-12_18-40-20](https://user-images.githubusercontent.com/202757/27044665-aa598b8a-4f9e-11e7-9b65-d2bfcce60d88.png)

tl;dr I'm not really sure what would lead to this specific scenario.

**Comment by tlhc** [comment]
may be OTA replace some files ?

**Comment by Frenzie** [comment]
> when I download the koreader-nightly-20170430, reinstall the koreader and attempt to OTA again，
koreader show already update to last version v2015.11-1054 , it is normal ?

You probably still have the latest OTA package in your /settings directory or something.

> may be OTA replace some files ?

Not sure what you mean? The OTA works by replacing files with their updated versions.

**Comment by tlhc** [comment]
nothing, I mean all of this is OTA causes, may be the OTA replace some files it shouldn't or not replace the old version files with their updated versions ?
when I download koreader-nightly-20170430 and reinstall and tap OTA, I think koreader will OTA to the last version (but I still have the last version),  but, not crash again. :)

**Comment by tlhc** [comment]
![img_3596 1](https://user-images.githubusercontent.com/2936748/27046707-05b385f6-4fd8-11e7-841c-af7f3ab4f407.JPG)

When I try to highlight the second paragraph In Linux .....（with link to other section）, it crash, but other text are ok. you can try :)

**Comment by Frenzie** [comment]
I can't reproduce a crash on that page (33 per the document's own count, 36 in the file) but oddly enough I can't select any text on that page, anywhere. It doesn't matter whether reflow is enabled or not. Very peculiar.

**Comment by Frenzie** [comment]
This may have been fixed by #3241. Could you please give it a try?

**Comment by robert00s** [comment]
In v2015.11-1426-gf0f59af problem still exist.

**Comment by Frenzie** [comment]
What about v2015.11-1435-g69ec6c0? It includes a pretty big change in #3547.

**Comment by robert00s** [comment]
No crash in v2015.11-1436-g15ec47d so we can close it now.

**Comment by Frenzie** [comment]
Thanks @TnS-hun ! ;-)

**Comment by khariche** [comment]
I am sorry to say I still experience the issu on a Cervantes 4.
Settings: pdf reflow, render quality default, without cropping and tried all zoom options
The only workaround I have found is to lower the "render quality" to low, but it is obviously not optimal.

Here is the crashlog
03/14/20-00:28:21 INFO  opening file /mnt/sd/agregation/sync/civilisation/Franquismo - memoria/CIVI-ESPAGNE-memoire-du-franquisme-cours-I-CNED.pdf
03/14/20-00:28:21 INFO  RD loaded plugin SSH at plugins/SSH.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin autosuspend at plugins/autosuspend.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin autoturn at plugins/autoturn.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin backgroundrunner at plugins/backgroundrunner.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin batterystat at plugins/batterystat.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin calibrecompanion at plugins/calibrecompanion.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin coverbrowser at plugins/coverbrowser.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin docsettingtweak at plugins/docsettingtweak.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin evernote at plugins/evernote.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin goodreads at plugins/goodreads.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin keepalive at plugins/keepalive.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin kosync at plugins/kosync.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin newsdownloader at plugins/newsdownloader.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin perceptionexpander at plugins/perceptionexpander.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin readtimer at plugins/readtimer.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin send2ebook at plugins/send2ebook.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin statistics at plugins/statistics.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin systemstat at plugins/systemstat.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin terminal at plugins/terminal.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin texteditor at plugins/texteditor.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin timesync at plugins/timesync.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin wallabag at plugins/wallabag.koplugin
03/14/20-00:28:21 INFO  RD loaded plugin zsync at plugins/zsync.koplugin
03/14/20-00:28:21 INFO  setting zoom mode to page
ffi.load: pthread
GET https://sync.koreader.rocks:443/syncs/progress/8094b5516c593858808d365c27aa5a74
host or service not provided, or not known
PUT https://sync.koreader.rocks:443/syncs/progress
host or service not provided, or not known
PUT https://sync.koreader.rocks:443/syncs/progress
host or service not provided, or not known
Error in pixWordMaskByDilation: imin = 1 is too small
Error in pixConnComp: pixs not defined
Error in boxaSelectBySize: boxas not defined
Error in boxaSelectBySize: boxas not defined
Error in pixaCreateFromBoxa: boxa not defined
Error in boxaSort2d: boxas not defined
Error in pixaSort2dByIndex: pixas not defined
Error in pixaaFlattenToPixa: paa not defined
Error in pixaGetBoxa: pixa not defined
./luajit: frontend/document/koptinterface.lua:785: attempt to get length of local 'boxes' (a nil value)
stack traceback:
	frontend/document/koptinterface.lua:785: in function 'getWordFromBoxes'
	frontend/document/koptinterface.lua:1016: in function 'getTextFromPositions'
	frontend/apps/reader/modules/readerhighlight.lua:705: in function 'handler'
	frontend/ui/widget/container/inputcontainer.lua:255: in function 'handleEvent'
	frontend/ui/uimanager.lua:736: in function 'sendEvent'
	frontend/ui/uimanager.lua:324: in function '__default__'
	frontend/ui/uimanager.lua:1076: in function 'handleInputEvent'
	frontend/ui/uimanager.lua:1137: in function 'handleInput'
	frontend/ui/uimanager.lua:1181: in function 'run'
	./reader.lua:298: in main chunk
	[C]: at 0x00013f05
